pkg/controller/master/master_controller.go

```diff
@@ -157,16 +182,114 @@ func (s *installStrategy) setupServiceAccountForJob(clusterNamespace string) (*k
 	return currentSA, nil
 }
 
+func (s *installStrategy) masterMachineSet(cluster *clustop.CombinedCluster) (*capi.MachineSet, error) {
+	masterMachineSetName := controller.MasterMachineSetName(cluster.Name)
+	return s.machineSetLister.MachineSets(cluster.Namespace).Get(masterMachineSetName)
+}
+
+func (s *installStrategy) availableMasterMachines(masterMachineSet *capi.MachineSet) ([]*capi.Machine, error) {
+	machines := []*capi.Machine{}
+	allMachines, err := s.machineLister.Machines(masterMachineSet.Namespace).List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+	for _, machine := range allMachines {
+		if !metav1.IsControlledBy(machine, masterMachineSet) ||
+			machine.DeletionTimestamp != nil {
+			continue
+		}
+		hasPublicName, err := hasPublicDNS(machine)
+		if err != nil {
+			return nil, fmt.Errorf("error checking whether machine %s/%s has public DNS: %v", machine.Namespace, machine.Name, err)
+		}
+		if !hasPublicName {
+			continue
+		}
+		machines = append(machines, machine)
+	}
+	return machines, nil
+}
+
+func (s *installStrategy) machinesReadyToInstall(masterMachineSet *capi.MachineSet) bool {
+	logger := logging.WithMachineSet(s.logger, masterMachineSet)
+	machines, err := s.availableMasterMachines(masterMachineSet)
+	if err != nil {
+		logger.Errorf("error determining available master machines: %v", err)
+		return false
+	}
+	if masterMachineSet.Spec.Replicas == nil {
+		logger.Errorf("no replicas defined for master machineset")
+		return false
+	}
+	// Only start an install if we have the number of replicas specified in the machineset
+	if len(machines) < int(*masterMachineSet.Spec.Replicas) {
+		return false
+	}
+	// Of the available machines, determine if any has not had its control plane installed
+	// If all of them have a control plane, then no need to run this job.
+	// TODO: determine how to handle upgrades, if they require a job.
+	for _, m := range machines {
+		if m.Status.Versions == nil || len(m.Status.Versions.ControlPlane) == 0 {
```

Comment:
We also need to clear `ControlPlane` in the actuator when the instance is created, to handle the case where the instance was deleted externally and had to be re-created.